msprime/simulations.py

```diff
@@ -157,14 +158,23 @@ def _replicate_generator(
     Generator function for the many-replicates case of the simulate
     function.
     """
-    # TODO We should encode the replicate index in here with the rest of the
-    # parameters. This will provide sufficient information to reproduce the
-    # simulation if necessary. Much simpler than encoding the details of
-    # the random number generator.
-    provenance_record = json.dumps(provenance_dict)
+
+    encoded_provenance = None
+    # The JSON is modified for each replicate to insert the replicate number.
+    # To avoid repeatedly encoding the same JSON (which can take milliseconds)
+    # we insert a replaceable string.
+    placeholder = '@@_MSPRIME_REPLICATE_INDEX_@@'
+    if provenance_dict is not None:
+        provenance_dict['parameters']['replicate_index'] = placeholder
+        encoded_provenance = provenance.json_encode_provenance(
+            provenance_dict, num_replicates)
+
     for j in range(num_replicates):
         sim.run(end_time)
-        tree_sequence = sim.get_tree_sequence(mutation_generator, provenance_record)
+        tree_sequence = sim.get_tree_sequence(
```

Comment:
I'm not following the logic here - do we ever use the string "Provenance was disabled"? Can we pass ``None`` if not? I found myself having to think about what this does a bit too much, how about:
```python
provenance_record = None
if encoded_provenance is not None:
       provenance_record = encoded_provenance.replace(f'"{placeholder}"', str(j))
```
It's a bit more verbose, but clearer? I guess a rule of thumb for conditional assignments is "don't use them if you have to split lines"?

Reply:
I agree the logic could be clearer. 
The intent here was that upon opening a file and looking at the provenance you clearly saw that it was disabled, rather than just getting "None", but I'm in danger of overthinking! I'll put it back to None.